### PR TITLE
Emit some additional information in the IDL/Snapshot to enable us to generate a java stub back from the IDL

### PR DESCRIPTION
**Pull request by karthikrg**

### CODE REVIEW

File: restli-server/src/main/java/com/linkedin/restli/internal/server/model/ResourceMethodDescriptor.java [hunk 1]
```diff
@@ -430,6 +437,85 @@ public String getMethodName()
...
+    if (unwrappedType instanceof ParameterizedType) {
+      return (Class<?>) ((ParameterizedType) unwrappedType).getRawType();
+    }
+    return (Class<?>) unwrappedType;
+  }
+
+  /**
+   * @return Resource Java method type.
+   */
+  public JavaMethodType getJavaMethodType()
```
Comment (mchen07) [inline restli-server/src/main/java/com/linkedin/restli/internal/server/model/ResourceMethodDescriptor.java:483]: ResourceMethodDescriptor has _interfaceType can be directly mapped to your JavaMethodType.

File: restli-server/src/main/java/com/linkedin/restli/internal/server/model/ResourceMethodDescriptor.java [hunk 2]
```diff
@@ -430,6 +437,85 @@ public String getMethodName()
...
+
+    switch (methodType) {
+      case ASYNC:
+        // Find the index of the callback param.
+        int callBackParamIndex = IntStream.range(0, _parameters.size())
+            .filter(i -> _parameters.get(i).getParamType() == Parameter.ParamType.CALLBACK)
+            .findFirst()
+            .orElse(-1);
+        // The actual param is wrapped inside the Callback as a generic param, so extract the embedded generic type.
+        return getFirstActualType((ParameterizedType) _method.getGenericParameterTypes()[callBackParamIndex]);
```
Comment (mchen07) [inline restli-server/src/main/java/com/linkedin/restli/internal/server/model/ResourceMethodDescriptor.java:455]: you should guard indexOutofRange, throw error if no callback parameter found.

File: restli-common/src/main/pegasus/com/linkedin/restli/restspec/RestMethodSchema.pdl [hunk 3]
```diff
@@ -10,6 +10,18 @@ record RestMethodSchema includes CustomAnnotationSchema, ServiceErrorsSchema, Su
...
+   * Java method type for this rest method
+   */
+  javaMethodType: optional JavaMethodType = "SYNC"
+
+  /**
+   * True if the java implementation of this method returns the result wrapper (aka CollectionResult for GET_ALL, or
+   * GetResult for GET or BatchResult for Batch_GET). False, if it returns the result as is. This is only applicable
+   * for read methods (GET/BATCH_GET/GET_ALL).
+   */
+  javaReturnsResultWrapper: optional boolean
```
Comment (mchen07) [inline restli-common/src/main/pegasus/com/linkedin/restli/restspec/RestMethodSchema.pdl:23]: For difference btw CreateResponse and CreateKVResponse, BatchCreateResult vs BatchCreateKVResult, are we planning to determine that based on returnEntity annotation?

Comment (karthikrg) [reply restli-common/src/main/pegasus/com/linkedin/restli/restspec/RestMethodSchema.pdl:23]: Correct. That is already in the IDL.

**Review by mchen07 — APPROVED** [review]

**Pull request merged**